modules/generator/processor/localblocks/processor.go

```diff
@@ -200,62 +204,105 @@ func (p *Processor) metricLoop() {
 }
 
 func (p *Processor) completeBlock() error {
-	p.blocksMtx.Lock()
-	defer p.blocksMtx.Unlock()
+
+	// Get a wal block
+	var firstWalBlock common.WALBlock
+	p.blocksMtx.RLock()
+	for _, e := range p.walBlocks {
+		firstWalBlock = e
+		break
+	}
+	p.blocksMtx.RUnlock()
+
+	if firstWalBlock == nil {
+		return nil
+	}
+
+	// Now create a new block
 
 	var (
 		ctx    = context.Background()
 		enc    = encoding.DefaultEncoding()
 		reader = backend.NewReader(p.wal.LocalBackend())
 		writer = backend.NewWriter(p.wal.LocalBackend())
 		cfg    = p.Cfg.Block
+		b      = firstWalBlock
 	)
 
-	for id, b := range p.walBlocks {
-
-		iter, err := b.Iterator()
-		if err != nil {
-			return err
-		}
-		defer iter.Close()
+	iter, err := b.Iterator()
+	if err != nil {
+		return err
+	}
+	defer iter.Close()
 
-		newMeta, err := enc.CreateBlock(ctx, cfg, b.BlockMeta(), iter, reader, writer)
-		if err != nil {
-			return err
-		}
+	newMeta, err := enc.CreateBlock(ctx, cfg, b.BlockMeta(), iter, reader, writer)
+	if err != nil {
+		return err
+	}
 
-		newBlock, err := enc.OpenBlock(newMeta, reader)
-		if err != nil {
-			return err
-		}
+	newBlock, err := enc.OpenBlock(newMeta, reader)
+	if err != nil {
+		return err
+	}
 
-		p.completeBlocks[newMeta.BlockID] = newBlock
+	// Add new block and delete old block
+	p.blocksMtx.Lock()
+	defer p.blocksMtx.Unlock()
 
-		err = b.Clear()
-		if err != nil {
-			return err
-		}
-		delete(p.walBlocks, id)
+	p.completeBlocks[newMeta.BlockID] = newBlock
 
-		// Only do 1 block per call
-		return nil
+	err = b.Clear()
+	if err != nil {
+		return err
 	}
+	delete(p.walBlocks, b.BlockMeta().BlockID)
 
 	return nil
 }
 
 func (p *Processor) GetMetrics(ctx context.Context, req *tempopb.SpanMetricsRequest) (*tempopb.SpanMetricsResponse, error) {
+	p.blocksMtx.RLock()
+	defer p.blocksMtx.RUnlock()
 
-	fetcher := traceql.NewSpansetFetcherWrapper(func(ctx context.Context, req traceql.FetchSpansRequest) (traceql.FetchSpansResponse, error) {
-		if p.headBlock == nil {
-			return traceql.FetchSpansResponse{}, fmt.Errorf("head block is nil")
+	// Blocks to check
+	blocks := make([]common.BackendBlock, 0, 1+len(p.walBlocks)+len(p.completeBlocks))
```

Comment:
This is going to create a lot of memory pressure with each read by copying memory around. We could loop through the blocks directly instead.

Reply:
Good point.  I wonder if we could keep this `blocks` object on the struct, since we're already locking, we could keep it updated on the iteration but not need to re-allocate new memory.

Reply:
I took this approach because it greatly simplifies the below logic and error handling, one loop instead of two or three (head block, wal blocks, complete blocks).  The overhead of this slice should be minimal - expect a few dozen blocks and single digit requests per second to this api.